src/Yaapii.Xambly/ThreadSafeCollection.cs

```diff
@@ -14,84 +17,100 @@ public class ThreadsafeCollection<T> : IList<T>, IList
         /// </summary>
-        private readonly List<T> items;
+        private readonly IScalar<List<T>> items;
 
         /// <summary>
         /// Lock object
         /// </summary>
-        private readonly object sync;
+        private readonly IScalar<object> sync;
 
         /// <summary>
         /// ctor
         /// </summary>
-        public ThreadsafeCollection()
-        {
-            this.items = new List<T>();
-            this.sync = new Object();
-        }
+        public ThreadsafeCollection() : this(
+            new Object(),
+            new EnumerableOf<T>()
+        )
+        { }
 
         /// <summary>
         /// ctor
         /// </summary>
-        /// <param name="syncRoot"></param>
-        public ThreadsafeCollection(object syncRoot)
-        {
-            if (syncRoot == null)
-                throw new ArgumentNullException("syncRoot");
-
-            this.items = new List<T>();
-            this.sync = syncRoot;
-        }
+        /// <param name="sync"></param>
+        public ThreadsafeCollection(object sync) : this(
+            new StickyScalar<object>(() =>
+            {
+                new FailNull(
+                    sync,
+                    "Syncronizing object is null"
+                ).Go();
+                return sync;
+            }),
+            new EnumerableOf<T>()
+        )
+        { }
 
         /// <summary>
         /// ctor
         /// </summary>
-        /// <param name="syncRoot">root object to sync</param>
+        /// <param name="sync">object to sync</param>
         /// <param name="list">list to instantiate from</param>
-        public ThreadsafeCollection(object syncRoot, IEnumerable<T> list)
-        {
-            if (syncRoot == null)
-                throw new ArgumentNullException("syncRoot");
-            if (list == null)
-                throw new ArgumentNullException("list");
-
-            this.items = new List<T>(list);
-            this.sync = syncRoot;
-        }
+        public ThreadsafeCollection(object sync, params T[] list) : this(
+            sync,
+            new EnumerableOf<T>(list)
+        )
+        { }
 
         /// <summary>
         /// ctor
         /// </summary>
-        /// <param name="syncRoot">root object to sync</param>
+        /// <param name="sync">object to sync</param>
         /// <param name="list">list to instantiate from</param>
-        public ThreadsafeCollection(object syncRoot, params T[] list)
-        {
-            if (syncRoot == null)
-                throw new ArgumentNullException("syncRoot");
-            if (list == null)
-                throw new ArgumentNullException("list");
-
-            this.items = new List<T>(list.Length);
-            for (int i = 0; i < list.Length; i++)
-                this.items.Add(list[i]);
+        public ThreadsafeCollection(object sync, IEnumerable<T> list) : this(
+            new StickyScalar<object>(() =>
+            {
+                new FailNull(
```

Comment:
restore the original exceptions or throw a specific one like ArgumentException.

Reply:
Throw ArgumentNullException